Feat: 내 장부 조회

ekzot1212 opened the pull request:
## Issue
- #14


## Overview
- 내 장부 조회 기능
- 장부 관련 파일명 변경

## 참고(optional)

ekzot1212 committed:
#14 Refactor: 장부의 파일명들을 Ledger 에서 AccountBook으로 변경

ekzot1212 committed:
Merge branch 'dev' into feature/accountBook

ekzot1212 committed:
Merge branch 'dev' into feature/accountBook

ekzot1212 committed:
Feat: 내 장부 조회 #14

ekzot1212 committed:
Refactor : 불필요한 어노테이션 제거 #14

dlwjddn123 commented on src/main/java/nbdream/accountBook/controller/AccountBookController.java:
```diff
@@ -0,0 +1,27 @@
+        GetAccountBookListResDto resDto = accountBookService.getMyAccountBookList(reqDto, memberId);
+        return ApiResponse.ok(resDto);
+    }
+}
```
GetAccountBookListReqDto를 @ModelAttribute 로 받고 있는데요 이는 형식이 맞지 않아요!
저희는 Json형태로 주고 받기 때문에 @RequestBody로 받아야 합니다.
이와 관련하여 참고하실 블로그 첨부드립니다!

https://tecoble.techcourse.co.kr/post/2021-05-11-requestbody-modelattribute/

dlwjddn123 commented on src/main/java/nbdream/accountBook/domain/AccountBook.java:
```diff
@@ -0,0 +1,49 @@
+        }
+        return -totalExpense;
+    }
+}
```
domain이 dto에 의존하는 방법은 좋지 않은 것 같아요! 특히 간단한 수정이 아닌 핵심 로직 같은 경우는!
AcountBookHistory에 의존하는 건 어떨까요? 

그리고 왠만하면 도메인에 존재하는 로직에 static은 어울리지 않습니다.
서비스에서 도메인을 직접 찾고 참조해야 할 것 같아요.

그리고 아주 사소한 거지만 if(!(dto.getExpense() == null)) -> if(dto.getExpense() != null) 이 더 가독성이 좋지 않을까요?

dlwjddn123 commented on src/main/java/nbdream/accountBook/domain/AccountBookCategory.java:
```diff
@@ -0,0 +1,35 @@
+                return category;
+            }
+        }
+        throw new IllegalArgumentException("Unknown category value: " + value);
```
enum을 전체적으로 잘 짜셨네요! IllegalArgumentException 요 에러를 CategoryNotFoundException으로 사용하지 않는 이유가 있을까요?

dlwjddn123 commented on src/main/java/nbdream/accountBook/domain/AccountBookHistory.java:
```diff
@@ -0,0 +1,52 @@
+        this.date = date;
+    }
+
+    public String getKoreanDayOfWeek() {
+        DayOfWeek dayOfWeek = this.date.getDayOfWeek();
+        return dayOfWeek.getDisplayName(TextStyle.FULL, Locale.KOREAN);
+    }
```
요일 처리 잘 해주신 거 같습니다!

dlwjddn123 commented on src/main/java/nbdream/accountBook/domain/Sort.java:
```diff
@@ -0,0 +1,26 @@
+                return sort;
+            }
+        }
+        throw new IllegalArgumentException("Unknown Sort value: " + value);
```
요 부분도 커스텀 에러로 처리하면 좋을 거 같아요

dlwjddn123 commented on src/main/java/nbdream/accountBook/domain/TransactionType.java:
```diff
@@ -0,0 +1,26 @@
+                return type;
+            }
+        }
+        throw new IllegalArgumentException("Unknown TransactionType value: " + value);
```
커스텀 에러!

dlwjddn123 commented on src/main/java/nbdream/accountBook/repository/specifications/AccountBookHistorySpecifications.java:
```diff
@@ -0,0 +1,70 @@
+            predicates.add(criteriaBuilder.equal(root.get("accountBook").get("id"), memberId));
+
+            // 카테고리
+            if (reqDto.getCategory() != null) {
+                try {
+                    AccountBookCategory categoryEnum = AccountBookCategory.fromValue(reqDto.getCategory());
+                    predicates.add(criteriaBuilder.equal(root.get("accountBookCategory"), categoryEnum));
+                } catch (IllegalArgumentException e) {
+
+                }
+            }
```
여기에서 try-catch문은 어떤 의미인가요?

ekzot1212 replied:
생각해보니 fromValue에서 예외처리 해놨기 때문에 별 의미가 없을거 같아서 빼겠습니다!

dlwjddn123 commented on src/main/java/nbdream/accountBook/repository/specifications/AccountBookHistorySpecifications.java:
```diff
@@ -0,0 +1,70 @@
+        return (root, query, criteriaBuilder) -> {
+            List<Predicate> predicates = new ArrayList<>();
+
+            predicates.add(criteriaBuilder.equal(root.get("accountBook").get("id"), memberId));
```
accountBook의 id와 memberId를 비교하는 게 맞나요??

ekzot1212 replied:
헉! 큰일날뻔 했네요 감사합니다

dlwjddn123 commented on src/main/java/nbdream/accountBook/repository/specifications/AccountBookHistorySpecifications.java:
```diff
@@ -0,0 +1,70 @@
+            }
+
+            // 수입/지출
+            if (reqDto.getCost() != null) {
+                try {
+                    TransactionType transactionTypeEnum = TransactionType.fromValue(reqDto.getCost());
+                    predicates.add(criteriaBuilder.equal(root.get("transactionType"), transactionTypeEnum));
+                } catch (IllegalArgumentException e) {
+
+                }
+            }
```
TransactionType transactionTypeEnum = TransactionType.fromValue(reqDto.getCost());
위의 부분에서 reqDto.getCost() -> reqDto.getTranscationType() 이 맞는 거 같네요

ekzot1212 replied:
api명세서에 보내주는 데이터가 cost :(expense,revenue) 라고 되어있더라구요
회의때 이름 바꾸는거 말해볼게요!

dlwjddn123 commented on src/main/java/nbdream/accountBook/service/AccountBookHistoryService.java:
```diff
@@ -0,0 +1,78 @@
+                .month(history.getDate().getMonthValue())
+                .day(history.getDate().getDayOfMonth())
+                .dayName(history.getKoreanDayOfWeek())
+                .expense(history.getTransactionType() == TransactionType.EXPENSE ? (long) history.getAmount() : null)
+                .revenue(history.getTransactionType() == TransactionType.REVENUE ? (long) history.getAmount() : null)
```
요 부분은 목요일 회의 때 request, response를 수정해서 지출 수입을 둘 다 받는게 아니라 카테고리? 타입?을 통해서 지출인지 수입인지 구분하도록 하자고 해야 합니다! 충희님도 기억해주세요!

ekzot1212 replied:
넵 혹시 그렇게 하면 어떤 장점이 있을까요??

dlwjddn123 commented on src/main/java/nbdream/accountBook/service/AccountBookService.java:
```diff
@@ -0,0 +1,52 @@
+
+    private final AccountBookRepository accountBookRepository;
+    private final AccountBookHistoryRepository accountBookHistoryRepository;
+    private final AccountBookHistoryService accountBookHistoryService;
```
AccountBookHistoryService 사실 요 서비스가 거의 장부 내역을 조회하는 로직 뿐이라 사실 이건 repository에서 처리해야 하는 부분이지 않을까요? 충희님 생각은 어떠신가요? Service끼리 의존하는 경우도 좋은 경우가 아니여서요. 순환 참조하는 경우도 조심해야 하고

dlwjddn123 commented on src/main/java/nbdream/accountBook/service/AccountBookService.java:
```diff
@@ -0,0 +1,52 @@
+        //금액 로직
+        Long totalRevenue = getTotalRevenue(items);
+        Long totalExpense = getTotalExpense(items);
+        Long totalCost = totalRevenue + totalExpense;
```
이 코드에서 getTotalRevenue를 호출 당하는 AccountBook이 memberId 값을 가진 Member가 맞나요?
memberId 값을 가진 AccountBook을 찾고, 그 객체에게 AccountBookHistory를 줘서 금액을 처리하는 게 더 좋을 것 같아요!

dlwjddn123 commented on src/main/java/nbdream/accountBook/service/dto/GetAccountBookListReqDto.java:
```diff
@@ -0,0 +1,26 @@
+    private String end;
+    private String cost;
+
+    private AccountBookCategory categoryEnum;
+    private TransactionType transactionTypeEnum;
+    private Sort sortEnum;
```
컨트롤러 쪽에서 요청할 때, @RequestBody의 Dto안에서Enum으로 바로 매핑이 되는 지 모르겠는데 혹시 테스트 해보셨을까요?

ekzot1212 replied:
해봤는데 String받아서 Enum타입으로 매핑은 안되더라구요

dlwjddn123 commented on src/main/java/nbdream/accountBook/service/dto/GetAccountBookResDto.java:
```diff
@@ -0,0 +1,25 @@
+import lombok.Builder;
+import lombok.Data;
+
+@Data
```
사소한 거지만 @Data는 @Setter를 가지고 있어서, 제 경우는 `@Getter`, `@NoArgsConstructor`, `@AllArgsConstructor` 를 주로 사용했던 거 같아요.

dlwjddn123 reviewed (commented):
장부 기능 구현하느라 고생 많으셨슴다.

ekzot1212 committed:
Refactor : 내 장부 조회 #14

- 컨트롤러 어노테이션 수정
- AccountBook 도메인이 dto가 아닌 AccountBookHistory를 의존하도록 수정
- 도메인 내부의 로직을 동적 메서드로 수정
- 커스텀 예외 처리
- 장부 내역 검색 시 id가 아닌 유저 객체로 매개변수 받게 수정
- Dto에서 @Data어노테이션 사용하지 않게 수정
